Add docs for JSON/Image Decoding on Main Thread profiling issues (#6513)
getsentry/sentry-docs: src/docs/product/issues/issue-details/performance-issues/json-decoding-main-thread/index.mdx

---
title: "JSON Decoding on Main Thread"
sidebar_order: 30
redirect_from:
  - /product/issues/performance-issues/json-decoding-main-thread/
description: "Learn more about JSON Decoding on Main Thread issues and how to diagnose and fix them."
---

> The main thread (or UI thread) in a mobile app is responsible for handling all user interaction, and maintaining a feeling of responsiveness requires the app to respond to gestures and taps in real time. When a long running operation blocks the main thread, the app becomes unresponsive, impacting the quality of the user experience.

Reviewer: ```suggestion
The main, (or UI) thread in a mobile app is responsible for handling all user interaction and needs to be able to respond to gestures and taps in real time. If a long-running operation blocks the main thread, the app becomes unresponsive, impacting the quality of the user experience.
```
Same comment as above. We can also write this if we want to be consistent with the File I/O on Main Thread language:
"The main UI thread in a mobile application handles user interface events such as button presses and page scrolls. To prevent App Hangs and Application Not Responding errors, the main UI thread should not be used for performing long-running operations. These kinds of operations block the whole UI until they finish running and get in the way of the user interacting with the app."

> Decoding a JSON blob into a data structure can be a long running operation, especially for large blobs. This issue type is created when we detect a long running JSON decoding operation on the main thread.

Reviewer: ```suggestion
The process of decoding a JSON blob into a data structure, is an example of a long-running operation that may impact app responsiveness. If a long-running JSON decoding operation is detected on the main thread, Sentry will flag it as an issue.

Decoding a JSON blob into a data structure can be a long running operation, especially for large blobs. This issue type is created when we detect a long running JSON decoding operation on the main thread.
```

## Detection Criteria

This issue is detected when profiling is enabled, and we detect a function related to JSON decoding running on the main thread for at least 40ms. This threshold was chosen so that we see the function appear in a sufficient number of samples collected by the profile (~10ms per sample x 4 = ~40ms total) before we detect the issue.

## Function Evidence

> On the issue details page, the **“Function Evidence”** section shows you more details about what caused the issue to be detected, including the following key information:
> 
> - **Transaction Name:** The name of the transaction that the issue was detected on
> - **Suspect function:** The function that triggered the issue detection (in this case, a JSON decoding function)
> - **Duration:** How long the function took to execute
> - **Sample count:** How many consecutive samples collected by the profiler contained the function
> 
> To view the entire profile associated with the issue, click the **“View Profile”** button:
> 
> ![JSON Decoding on Main Thread Function Evidence](json-decoding-function-evidence.png)
> 
> The profile will indicate where the suspect function was called from and the other functions that are called *by* the suspect function:
> 
> ![JSON Decoding on Main Thread Profile](json-decoding-profile.png)

Reviewer: ```suggestion
You can find additional information about your JSON Decoding on Main Thread problem by looking at four main aspects in the "Function Evidence" section:

- **Transaction Name:** The name of the transaction where the issue was detected.
- **Suspect function:** The function that triggered the issue detection (in this case, a JSON decoding function).
- **Duration:** How long the function took to execute.
- **Sample count:** The number of consecutive samples collected by the profiler that contained the function.

View it by going to the **Issues** page in Sentry, clicking on the issue you want to examine, then scrolling down to the "Function Evidence" section in the "Details" tab. To view the entire profile associated with the issue, click the “View Profile” button:

![JSON Decoding on Main Thread Function Evidence](json-decoding-function-evidence.png)

The profile will indicate where the suspect function was called from, along with other functions being called *by* the suspect function:

![JSON Decoding on Main Thread Profile](json-decoding-profile.png)
```
